### ISSUE TITLE
Create AJAX function and triggers and handling of progress indicator for the user

### ISSUE THREAD

**Issue by mehulkaklotar** [10up/plugin-check#34]
When the user triggers the checks by clicking on the “Check it” button, an admin AJAX request should be triggered to initiate the checks, and each check will run sequentially with each request in limited parameters. The response to the AJAX request will be to output checks with their respective results into the table format below the page options.

The AJAX callback function should always check for the nonces to verify the requests and should go ahead according to the verification.

There should be a visible indication of requests being made on the page for the user to see that the checks are in progress. After the successful completion of all checks, there should be an indication for the user that checks are complete and results are populated into the table. When the request has an error when running checks, it should indicate to the user that there has been some error while running the checks.

## Acceptance Criteria

- Class `Admin_AJAX` file should exist at `includes/Admin`
- Take output of the `add_management_page()` function available in the `add_page()` method of the `Admin_Page` class in to `$hook` variable.  (**Admin_Page** class introduced in https://github.com/10up/plugin-check/issues/32)
- Add hook on `load-{$hook}` and attach the `initialize_page` method to it in the `Admin_Page` class.
- `initialize_page()` method will have the `admin_enqueue_scripts` hook and callback method will be `enqueue_scripts`.
- Register new protected variable called `$admin_ajax` in the `Admin_Page` class.
- Add `constructor()` method in the `Admin_Page` class, and instantiate `Admin_AJAX` class in that.
- Add the `get_nonce()` method in the `Admin_AJAX` class. It will return the nonce, which can be used in the `Admin_Page` class.
- `Admin_Page` class should have the `enqueue_scripts()` method, which will register a new custom JS with the name `plugin-check-admin.js`, and this JS file will be placed at the top level of the plugin's directory at `plugin-check/assets`.
- The `plugin-check-admin.js` script should expose two PHP variables: the `admin-ajax` URL and `nonce` (which we get from `Admin_AJAX->get_nonce()`).
- `Admin_AJAX` will have `add_hooks()` method.
- `Admin_AJAX` will have `run_check()` method.
- `Admin_AJAX->add_hooks()` method will hook the ajax call with the `run_check()` method for only logged-in users.
- When the admin clicks on the `Check it!` button (button details available at https://github.com/10up/plugin-check/issues/32), the below process should start:
    1. JS sends an AJAX `POST` request to `admin-ajax.php` with the `plugin_check_run_check` action.
    2. `plugin_check_run_check` action will call the `run_check()` method.
        - The `run_check()` method gets the variables as below:
             - `nonce` - Basically, it is nonce to verify it with the `check_ajax_referer()` function.
        - The `run_check()` method will now respond to the success message if the nonce is valid else, return an error message in the JSON format.
    3. JS will get the details of the success or error messages and display them in the console.

## Test Coverage

**Admin_AJAX** class
- Add a test for the `Admin_Page->initialize_page()` method to check JS script only loads on the plugin-check.php only.
- Add a test for the `Admin_AJAX->add_hooks()` method.
- Add a test case for the `Admin_AJAX->get_nonce()` method.
- Add a test for the `Admin_AJAX->run_check()` method.

**Admin_Page** class
- Verified that the `plugin-check-admin.js` script is enqueued only on the `plugin-check` page.

**Comment by vishalkakadiya** [comment]
@jjgrainger Updated information in this epic. Can you review the details here, thank you!

**Comment by jjgrainger** [comment]
Thanks @vishalkakadiya 

I think we need to update this to include much more detail as to how this would work. I've included some points to consider:

- We will need to use JavaScript here to handle sending the AJAX request to the `admin-ajax.php` endpoint. This JS will be responsible for sending across the plugin and the checks we want to run based on the form submission on the tools page.
- Something to consider is that due to the limitations we will need to run each check individually. So the JS on this page will likely send multiple requests, one after another, for example:
    - User hits the 'Check it' button.
    - JS sends an AJAX request to `admin-ajax.php` with the plugin to check.
    - The server-side admin AJAX hook will run preparations, run the first check and return the results. It will also return the state of the checks, so the JS will no the next request to send to run the next check.
    - This process continues until the last check is run and the final results returned.
    - Then the results are displayed on the page (#35)
- The above is important and will be something we need to add a lot more detail on how this would work. 
- The `AJAX_Runner` class wouldn't be extended but may contain some methods to handle running checks and potentially the current state.
- The other important element to this is how we indicate the current progress to the user.

Let me know if you have any questions, happy to jump on a call to go over this.

**Comment by vishalkakadiya** [comment]
@jjgrainger Updated the description now with more details. Let me know if still need to add anything. Thank you!

**Comment by vishalkakadiya** [comment]
@felixarntz Can you also take a quick look at the description of this task? Thank you! 🙂

**Comment by jjgrainger** [comment]
Thanks @vishalkakadiya 

Great start but I think we need to iterate on a couple of things:

- Let's update `initialize_hooks()` to `add_hooks()` for consistency across the project.
- Additionally, I think we should instantiate and call all classes with `add_hooks()` methods in the [`Plugin_Main` file here](https://github.com/10up/plugin-check/blob/trunk/includes/Plugin_Main.php#L54) rather than in their own class constructors.
- We shouldn't use the `wp_ajax_nopriv_{$action}` as we would want these requests to come from a logged in user.
- Lets use the full plugin name for the action `wp_ajax_plugin_check_run_checks`.
- Lets include what data will be passed by the AJAX request.
- We should include nonce verification for these requests with `check_ajax_referer` also.
- Let's not place the `plugin-check-admin.js` in the includes folder but have an top-level assets folder for any JS or CSS needed.

Also, there are a lot of methods I don't think we'll need (`get_checks`, `check_prepare` `check_cleanup`). A lot of the heavy lifting for preparing and running checks will be handled by the #15 so this class could be simplified to handling the AJAX request by verifying the nonce, calling the `run()` method on the `AJAX_Runner`, and then return the results in the response.

Happy to jump on a call to talk through this if you have any questions.

Thanks

**Comment by vishalkakadiya** [comment]
@jjgrainger I have updated the description according to your feedback, and it is ready for your following review. Thank you! 🙂

**Comment by felixarntz** [comment]
@vishalkakadiya I believe, before we can finalize the ACs here, we need to finalize the ACs on #42, and the work in #15 is a dependency at least for the implementation as well.

Leaving some early feedback here:
* Should we call this class `Admin_AJAX` instead of `Admin_Checks`?
* We should not just use `admin_enqueue_scripts` and enqueue the script on all WP Admin pages, as that is a bad practice. It would add the script everywhere although it's only needed on our own admin page. Instead, I think the logic to enqueue the script should happen in the `Admin_Page` class. You can use the `$hook` returned by the `add_management_page()` call to add an action to `load-{$hook}`, e.g. add a method like `initialize_page()`. That method will then only be run if that page is the current admin page. So if you then add an `admin_enqueue_scripts` callback, it will only enqueue the script on our own admin page.
* I think it would be a good idea to instantiate the `Admin_AJAX` class within the `Admin_Page` class constructor, so that the admin page class can control its execution and also later access certain data from it (for example the nonce we need here should be controlled/provided by the `Admin_AJAX` class).
* You could add a method `get_nonce()` to the `Admin_AJAX` class that returns that nonce, so that the `Admin_Page` class can use it when providing data to the script.
* We should use `wp_add_inline_script()` to provide data from PHP to the script, not `wp_localize_script()`. The latter is intended to provide i18n strings to a script.

Regarding the exact JS logic, I think that is a bit preliminary here and has several limitations. We cannot just send a single AJAX request to run checks, we also need additional requests to run the global preparation and cleanup step (see #42). We also probably need one AJAX call to get the available checks, so that we can cycle through them and pass them to the AJAX request for running a single check.

I think we also need to think further how we want to present the output etc. I think this issue may be a bit much for one issue, so maybe we can break it down into smaller pieces. Since a lot of this depends on #42, we could unblock this issue now if we decide to just start scaffolding the infrastructure pieces needed here. How about we just focus on the following:
* Enqueue the script.
* Add the script file (for now just some `console.log( 'Success' )` or something like that.
* Scaffold a single AJAX handler for now, e.g. `wp_ajax_plugin_check_run_check`. It can just check the nonce for now and output some dummy JSON response like `{ success: true|false }` or something like that.

A separate issue could then be focused on actually implementing the AJAX requests, and yet another issue to implement the script.
cc @jjgrainger

**Comment by vishalkakadiya** [comment]
@felixarntz I liked the idea of separating the actual implementation then this issue, so I have simplified this issue now as per your feedback, and it is ready for review. Thank you! 🙂

**Comment by vishalkakadiya** [comment]
@felixarntz I have opened the new issue #84, but I need the starting point of it to write AC of that. Can you briefly explain how we will utilize the `AJAX_Runner` class methods with the `Admin_Page` and `Admin_AJAX` classes? Thank you!

CC @jjgrainger If you can give some input here, thanks!

**Comment by felixarntz** [comment]
@vishalkakadiya The current ACs here look almost good to go, but you're still referring to an AJAX action / method `run_checks`. This is something that we will not have, as we have already decided that every check will be run via its own AJAX request, as otherwise the request would likely take too long to respond as the number of checks grows.

Also you're still mentioning a bunch of JS logic in this issue; we don't know yet how exactly this should work as it's still blocked by the other issues, and also having just one AJAX endpoint is not going to be enough for the JS script to work.

As mentioned in my last comment, I think at this point we should only scaffold a basic `run_check` AJAX action / method that returns a dummy JSON response, and its purpose should eventually be to run a _single_ check. We should furthermore just keep the JS script do nothing (e.g. just `console.log` something) because we still lack the infrastructure for what it needs to do.

**Comment by felixarntz** [comment]
@vishalkakadiya 
> I have opened the new issue #84, but I need the starting point of it to write AC of that. Can you briefly explain how we will utilize the `AJAX_Runner` class methods with the `Admin_Page` and `Admin_AJAX` classes? Thank you!

I replied to this as part of https://github.com/10up/plugin-check/issues/84#issuecomment-1439373450. Note that I think we will only need to access the `AJAX_Runner` in the `Admin_AJAX` class - I don't think the `Admin_Page` class needs to do anything with it.

**Comment by vishalkakadiya** [comment]
@felixarntz I have updated AC according to your last feedback removed extra things, and it is neat and clean now. Thank you!

**Comment by felixarntz** [comment]
@vishalkakadiya Thank you, ACs LGTM!

**Comment by vishalkakadiya** [comment]
Thank you! @felixarntz @jjgrainger

**Comment by vishalkakadiya** [comment]
@jjgrainger Meanwhile, my PRs are in review. I’m starting on [#34](https://github.com/10up/plugin-check/issues/34).
CC @eclarke1

**Comment by vishalkakadiya** [comment]
@felixarntz @jjgrainger I have raised [PR for this](https://github.com/10up/plugin-check/pull/91). Thank you!
